ArduPlane/quadplane.cpp

```diff
@@ -3219,51 +3221,27 @@ float QuadPlane::get_weathervane_yaw_rate_cds(void)
     /*
       we only do weathervaning in modes where we are doing VTOL
       position control. We also don't do it if the pilot has given any
-      yaw input in the last 3 seconds.
+      yaw input in the last 2 seconds.
     */
     if (!in_vtol_mode() ||
         !motors->armed() ||
-        weathervane.gain <= 0 ||
         plane.control_mode == &plane.mode_qstabilize ||
 #if QAUTOTUNE_ENABLED
         plane.control_mode == &plane.mode_qautotune ||
 #endif
         plane.control_mode == &plane.mode_qhover
         ) {
-        weathervane.last_output = 0;
-        return 0;
-    }
-    const uint32_t tnow = millis();
-    if (plane.channel_rudder->get_control_in() != 0) {
-        weathervane.last_pilot_input_ms = tnow;
-        weathervane.last_output = 0;
-        return 0;
-    }
-    if (tnow - weathervane.last_pilot_input_ms < 3000) {
-        weathervane.last_output = 0;
-        return 0;
+        return 0.0;
     }
 
-    float roll = wp_nav->get_roll() / 100.0f;
-    if (fabsf(roll) < weathervane.min_roll) {
-        weathervane.last_output = 0;
-        return 0;        
-    }
-    if (roll > 0) {
-        roll -= weathervane.min_roll;
-    } else {
-        roll += weathervane.min_roll;
-    }
-    
-    float output = constrain_float((roll/45.0f) * weathervane.gain, -1, 1);
-    if (should_relax()) {
-        output = 0;
+    // relax weathervane controller if we think we have landed
+    weathervane->set_relax(should_relax());
+
+    if (weathervane->should_weathervane(wp_nav->get_roll(), wp_nav->get_pitch(), plane.channel_rudder->get_control_in(), plane.relative_ground_altitude(plane.g.rangefinder_landing))) {
```

Comment:
This is a existing issue, but we should probably be asking the attitude or position controller rather than WP nav. No need to do it in this PR.